In [1]:
import numpy as np
import pandas as pd
import csv
import matplotlib.pyplot as plt

In [2]:
# Specify the path to your Excel file
csv_file_path = "C:\\Users\\Mihir Shah\\Downloads\\penguins_lter.csv"

# Read the Excel file into a DataFrame
df = pd.read_csv(csv_file_path)

# Now you can work with the DataFrame, for example, printing the first few rows
df

studyName  Sample Number                              Species  Region  \
0     PAL0708              1  Adelie Penguin (Pygoscelis adeliae)  Anvers   
1     PAL0708              2  Adelie Penguin (Pygoscelis adeliae)  Anvers   
2     PAL0708              3  Adelie Penguin (Pygoscelis adeliae)  Anvers   
3     PAL0708              4  Adelie Penguin (Pygoscelis adeliae)  Anvers   
4     PAL0708              5  Adelie Penguin (Pygoscelis adeliae)  Anvers   
..        ...            ...                                  ...     ...   
339   PAL0910            120    Gentoo penguin (Pygoscelis papua)  Anvers   
340   PAL0910            121    Gentoo penguin (Pygoscelis papua)  Anvers   
341   PAL0910            122    Gentoo penguin (Pygoscelis papua)  Anvers   
342   PAL0910            123    Gentoo penguin (Pygoscelis papua)  Anvers   
343   PAL0910            124    Gentoo penguin (Pygoscelis papua)  Anvers   

        Island               Stage Individual ID Clutch Completion  Date Egg  \
0    Torgersen  Adult, 1 Egg Stage          N1A1               Yes  11/11/07   
1    Torgersen  Adult, 1 Egg Stage          N1A2               Yes  11/11/07   
2    Torgersen  Adult, 1 Egg Stage          N2A1               Yes  11/16/07   
3    Torgersen  Adult, 1 Egg Stage          N2A2               Yes  11/16/07   
4    Torgersen  Adult, 1 Egg Stage          N3A1               Yes  11/16/07   
..         ...                 ...           ...               ...       ...   
339     Biscoe  Adult, 1 Egg Stage         N38A2                No   12/1/09   
340     Biscoe  Adult, 1 Egg Stage         N39A1               Yes  11/22/09   
341     Biscoe  Adult, 1 Egg Stage         N39A2               Yes  11/22/09   
342     Biscoe  Adult, 1 Egg Stage         N43A1               Yes  11/22/09   
343     Biscoe  Adult, 1 Egg Stage         N43A2               Yes  11/22/09   

     Culmen Length (mm)  Culmen Depth (mm)  Flipper Length (mm)  \
0                  39.1               18.7                181.0   
1                  39.5               17.4                186.0   
2                  40.3               18.0                195.0   
3                   NaN                NaN                  NaN   
4                  36.7               19.3                193.0   
..                  ...                ...                  ...   
339                 NaN                NaN                  NaN   
340                46.8               14.3                215.0   
341                50.4               15.7                222.0   
342                45.2               14.8                212.0   
343                49.9               16.1                213.0   

     Body Mass (g)     Sex  Delta 15 N (o/oo)  Delta 13 C (o/oo)  \
0           3750.0    MALE                NaN                NaN   
1           3800.0  FEMALE            8.94956          -24.69454   
2           3250.0  FEMALE            8.36821          -25.33302   
3              NaN     NaN                NaN                NaN   
4           3450.0  FEMALE            8.76651          -25.32426   
..             ...     ...                ...                ...   
339            NaN     NaN                NaN                NaN   
340         4850.0  FEMALE            8.41151          -26.13832   
341         5750.0    MALE            8.30166          -26.04117   
342         5200.0  FEMALE            8.24246          -26.11969   
343         5400.0    MALE            8.36390          -26.15531   

                           Comments  
0    Not enough blood for isotopes.  
1                               NaN  
2                               NaN  
3                Adult not sampled.  
4                               NaN  
..                              ...  
339                             NaN  
340                             NaN  
341                             NaN  
342                             NaN  
343                             NaN  

[344 rows x 17 columns]

Removing All Missing & Nan Values from the Dataset to Run Logistic Regression

In [3]:
# Drop rows with any NaN values
df.dropna(axis=0, how='any', inplace=True)

Applying Logistic Regression to the Dataset & Loading Packages of Sklearn

In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Define features (X) and target variable (y)

In [5]:
X = df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)']]
y = df['Sex'] 

Splitting the data into train and test sets (80:20 split)

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Initializing and fitting the logistic regression model

In [7]:
logistic_model = LogisticRegression()
logistic_model.fit(X_train, y_train)

LogisticRegression()

Predicting on the test set

In [8]:
y_pred = logistic_model.predict(X_test)

Calculating the accuracy of the model

In [9]:
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

Accuracy: 0.6666666666666666


2.Decision Tree Classifier

Loading Necessary Packages

In [10]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

Splitting the Datasets

In [11]:
# Split the DataFrame into training and testing sets
X = df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df['Sex']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

Fitting the Classifier

In [12]:
# Create and fit the Decision Tree Classifier
clf = DecisionTreeClassifier()
clf.fit(X_train, y_train)

DecisionTreeClassifier()

Evaluate the Classifier & finding the Accuracy

In [14]:
from sklearn.metrics import accuracy_score

y_pred = clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)

Accuracy: 0.6666666666666666


3.Random Forrest Classifier

Importing necessary libraries

In [15]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report

Split the data into training and testing sets (80:20)

In [16]:
X = df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df['Sex']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

Train the Random Forest Classifier

In [17]:
rf_clf = RandomForestClassifier()
rf_clf.fit(X_train, y_train)

RandomForestClassifier()

Evaluate the Random Forest Classifier

In [18]:
y_rf_pred = rf_clf.predict(X_test)
rf_accuracy = accuracy_score(y_test, y_rf_pred)
print("Random Forest Classifier Accuracy:", round(rf_accuracy,2))

Random Forest Classifier Accuracy: 0.67


4.Naive Bayes

Importing Necessary Libraries

In [19]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

Split the Dataset and Defining the Variables

In [20]:
X = df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df['Sex']

# Split the data into training and testing sets (80:20)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

Training the Naive Bayes Classifier

In [21]:
nb_clf = GaussianNB()
nb_clf.fit(X_train, y_train)

GaussianNB()

Evaluating the Naive Bayes Classifier

In [22]:
y_nb_pred = nb_clf.predict(X_test)
nb_accuracy = accuracy_score(y_test, y_nb_pred)
print("Naive Bayes Classifier Accuracy:", round(nb_accuracy, 2))

Naive Bayes Classifier Accuracy: 0.67


5. Ada Boost Classifier

Loading Necessary Libraries

In [23]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score

Separate features and target variable

In [24]:
X = df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df['Sex']

# Split the data into training and testing sets (80:20)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

Initialize and train the AdaBoost classifier

In [25]:
ada_boost_clf = AdaBoostClassifier()
ada_boost_clf.fit(X_train, y_train)

AdaBoostClassifier()

Evaluate the AdaBoost classifier

In [26]:
y_ada_boost_pred = ada_boost_clf.predict(X_test)
ada_boost_accuracy = accuracy_score(y_test, y_ada_boost_pred)
print("AdaBoost Classifier Accuracy:", round(ada_boost_accuracy, 2))

AdaBoost Classifier Accuracy: 1.0


Classifiers and their Accuracy(%)

Logistic Regression - 66%

Decision Tree - 66%

Random Forrest - 67%

Naive Bayes - 67%

Ada Boost - 100%

Now the Models of Bootstrapping, K-fold Cross Validation & Cross Validation with LOO with this 5 Classifiers one by one

Bootstrapping

1. Bootstrapping with Logistic Regression

In [41]:
import pandas as pd
from sklearn.model_selection import train_test_split, StratifiedShuffleSplit
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Subsample the data to 200 samples
df_boot = df.sample(200, replace=True)

# Split the subsampled data into training and testing sets
X = df_boot[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df_boot['Sex']

# Use stratified shuffle split to ensure data balance
sss = StratifiedShuffleSplit(n_splits=1, test_size=0.2)

# Perform bootstrapping with a single iteration
train_index, test_index = next(iter(sss.split(X, y)))
X_train, X_test = X.iloc[train_index], X.iloc[test_index]
y_train, y_test = y.iloc[train_index], y.iloc[test_index]

# Create the Logistic Regression model
model = LogisticRegression()

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = model.predict(X_test)

# Calculate the accuracy score
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)

Accuracy: 0.725


2.Bootstrapping with Decision Tree

In [42]:
import pandas as pd
from sklearn.model_selection import train_test_split, StratifiedShuffleSplit
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Subsample the data to 200 samples
df_boot = df.sample(200, replace=True)

# Split the subsampled data into training and testing sets
X = df_boot[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df_boot['Sex']

# Use stratified shuffle split to ensure data balance
sss = StratifiedShuffleSplit(n_splits=1, test_size=0.2)

# Perform bootstrapping with a single iteration
train_index, test_index = next(iter(sss.split(X, y)))
X_train, X_test = X.iloc[train_index], X.iloc[test_index]
y_train, y_test = y.iloc[train_index], y.iloc[test_index]

# Create the Decision Tree model
model = DecisionTreeClassifier()

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = model.predict(X_test)

# Calculate the accuracy score
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)

Accuracy: 1.0


3.Bootstrapping with Random Forrest

In [43]:
import pandas as pd
from sklearn.model_selection import train_test_split, StratifiedShuffleSplit
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Subsample the data to 200 samples
df_boot = df.sample(200, replace=True)

# Split the subsampled data into training and testing sets
X = df_boot[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df_boot['Sex']

# Use stratified shuffle split to ensure data balance
sss = StratifiedShuffleSplit(n_splits=1, test_size=0.2)

# Perform bootstrapping with a single iteration
train_index, test_index = next(iter(sss.split(X, y)))
X_train, X_test = X.iloc[train_index], X.iloc[test_index]
y_train, y_test = y.iloc[train_index], y.iloc[test_index]

# Create the Random Forest model
model = RandomForestClassifier()

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = model.predict(X_test)

# Calculate the accuracy score
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)

Accuracy: 1.0


4.Bootstrapping with Naive Bayes

In [44]:
import pandas as pd
from sklearn.model_selection import train_test_split, StratifiedShuffleSplit
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# Subsample the data to 200 samples
df_boot = df.sample(200, replace=True)

# Split the subsampled data into training and testing sets
X = df_boot[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df_boot['Sex']

# Use stratified shuffle split to ensure data balance
sss = StratifiedShuffleSplit(n_splits=1, test_size=0.2)

# Perform bootstrapping with a single iteration
train_index, test_index = next(iter(sss.split(X, y)))
X_train, X_test = X.iloc[train_index], X.iloc[test_index]
y_train, y_test = y.iloc[train_index], y.iloc[test_index]

# Create the Naive Bayes model
model = GaussianNB()

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = model.predict(X_test)

# Calculate the accuracy score
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)

Accuracy: 1.0


5.Bootstrapping with Ada-Boost

In [45]:
import pandas as pd
from sklearn.model_selection import train_test_split, StratifiedShuffleSplit
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score

# Subsample the data to 200 samples
df_boot = df.sample(200, replace=True)

# Split the subsampled data into training and testing sets
X = df_boot[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']]
y = df_boot['Sex']

# Use stratified shuffle split to ensure data balance
sss = StratifiedShuffleSplit(n_splits=1, test_size=0.2)

# Perform bootstrapping with a single iteration
train_index, test_index = next(iter(sss.split(X, y)))
X_train, X_test = X.iloc[train_index], X.iloc[test_index]
y_train, y_test = y.iloc[train_index], y.iloc[test_index]

# Create the AdaBoost Classifier model
model = AdaBoostClassifier()

# Fit the model to the training data
model.fit(X_train, y_train)

# Make predictions on the testing data
y_pred = model.predict(X_test)

# Calculate the accuracy score
accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', accuracy)

Accuracy: 1.0


It is observed that, here except Logistic Regression Classifier the Accuracy for other classifiers are 100% but 72.5% in case of Logistic Regression. So,it is better to not apply Logistic Regression Classifier for bootstrapping in this Dataset to predict Sex of Penguin and apply any of 4 Classifiers.

(2) K-fold Cross Validation

1. K-fold Cross Validation with Logistic Regression

In [52]:
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Define the K-fold cross-validation object
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Create the Logistic Regression model
model = LogisticRegression()

# Perform 5-fold cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=kf, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 0.9349999999999999


C:\Users\Mihir Shah\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\Mihir Shah\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression

2. K-fold Cross Validation with Decision Tree

In [48]:
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Define the K-fold cross-validation object
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Create the Decision Tree model
model = DecisionTreeClassifier()

# Perform 5-fold cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=kf, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


3. K-fold Cross Validation with Random Forrest

In [49]:
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Define the K-fold cross-validation object
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Create the Random Forest model
model = RandomForestClassifier()

# Perform 5-fold cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=kf, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


4. K-fold Cross Validation with Naive bayes

In [50]:
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# Define the K-fold cross-validation object
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Create the Naive Bayes model
model = GaussianNB()

# Perform 5-fold cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=kf, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


5. K-fold Cross Validation with Ada Boost

In [51]:
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score

# Define the K-fold cross-validation object
kf = KFold(n_splits=5, shuffle=True, random_state=42)

# Create the AdaBoost Classifier model
model = AdaBoostClassifier()

# Perform 5-fold cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=kf, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


Here, it is observed that each Classifier is having 100% Accuracy except logistic Regression Classifier as it has 93% Accuracy in Cross Validation so any of 4 Classifier will be appropriate in K-fold Cross Validation Model.

(3) Cross-Validation with Leave One Out(LOO)

1. LOOCV with Logistic Regression

In [58]:
import pandas as pd
from sklearn.model_selection import LeaveOneOut, cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import StandardScaler

# Preprocess the data by standardizing the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']])

# Define the Leave-One-Out cross-validation object
loo = LeaveOneOut()

# Create the Logistic Regression model with increased max_iter
model = LogisticRegression(max_iter=1000)

# Perform Leave-One-Out cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, X_scaled, df['Sex'], cv=loo, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


2. LOOCV with Decision Tree

In [54]:
import pandas as pd
from sklearn.model_selection import LeaveOneOut, cross_val_score
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Define the Leave-One-Out cross-validation object
loo = LeaveOneOut()

# Create the Decision Tree model
model = DecisionTreeClassifier()

# Perform Leave-One-Out cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=loo, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


3. LOOCV with Random Forrest

In [55]:
import pandas as pd
from sklearn.model_selection import LeaveOneOut, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Define the Leave-One-Out cross-validation object
loo = LeaveOneOut()

# Create the Random Forest model
model = RandomForestClassifier()

# Perform Leave-One-Out cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=loo, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


4. LOOCV with Naive-bayes

In [56]:
import pandas as pd
from sklearn.model_selection import LeaveOneOut, cross_val_score
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# Define the Leave-One-Out cross-validation object
loo = LeaveOneOut()

# Create the Naive Bayes model
model = GaussianNB()

# Perform Leave-One-Out cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=loo, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


5. LOOCV with Ada Boost

In [57]:
import pandas as pd
from sklearn.model_selection import LeaveOneOut, cross_val_score
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import accuracy_score

# Define the Leave-One-Out cross-validation object
loo = LeaveOneOut()

# Create the AdaBoost Classifier model
model = AdaBoostClassifier()

# Perform Leave-One-Out cross-validation and calculate the accuracy scores
accuracy_scores = cross_val_score(model, df[['Culmen Length (mm)', 'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)']], df['Sex'], cv=loo, scoring='accuracy')

# Print the average accuracy score
print("Average Accuracy:", accuracy_scores.mean())

Average Accuracy: 1.0


Here, it is observed that All 5 Classifiers in LOO Cross validation is giving 100% Accuracy so all can be appllied in the process of predicting Sex of Penguin.

Overall Summarry

In the Palmer-penguin's Dataset where we have to predict the Sex of Penguin.

Here, first we run all 5 classifiers individually which are Logistic Regression, Decision Tree,Random Forrest, Naive Bayes and Ada-Boost. Among them, Ada-Boost Classifier has given 100% Accuracy, Naive Bayes and Random Forrest has given 67% and Logistic Regression and Decision Tree has given 66% Accuracy.

So, for Individual Case, Ada-Boost Classifier is best classifier to apply for prediction of Sex of Penguin.

For, Bootstrapping it is observed that except Logistic Regression Classifier the Accuracy for other classifiers are 100% but 72.5% in case of Logistic Regression.

So,it is better to not apply Logistic Regression Classifier for bootstrapping in this Dataset to predict Sex of Penguin and apply any of 4 Classifiers.

For K-fold Cross validation, it is observed that each Classifier is having 100% Accuracy except logistic Regression Classifier as it has 93% Accuracy in Cross Validation so any of 4 Classifier will be appropriate in K-fold Cross Validation Model.

For LOOCV,it is observed that All 5 Classifiers in LOO Cross validation is giving 100% Accuracy so all can be appllied in the process of predicting Sex of Penguin.

LOOCV is the only Model out of 3 where all 5 Classifiers are giving it's full 100% Accuracy.

In the End, it is undeniable fact that Ada-Boost will work best in case of Individual Classifiers and LOO Cross Validation Models will give 100% Accuracy with application of all 5 Classifiers. And, if we are applying Bootstrapping or K-fold Cross validation Models then 4 Classifiers out of 5 show results of 100% Accuracy except Logistic Regresssion in Predicting Sex of Penguin.
So,application of  Ada-Boost Classifier with any of 3 Models Bootstrapping, K-fold Cross validation and Leave One Out Crossvalidation will work best and give appropriate and more accurate results in predicting Sex of Penguin.